ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/PostCommitValidation.scala

```diff
@@ -0,0 +1,230 @@
+  *
+  * Post-commit validation is relevant for two reasons:
+  * - keys can be referenced by two concurrent interpretations, potentially leading to
+  *   either create nodes with duplicate active keys or lookup-by-key nodes referring to
+  *   inactive keys
+  * - the transaction's ledger effective time is determined after interpretation,
+  *   meaning that causal monotonicity cannot be verified while interpreting a command
+  */
+sealed trait PostCommitValidation {
+
+  def validate(
+      transaction: Transaction,
+      transactionLedgerEffectiveTime: Instant,
+      divulged: Set[ContractId],
+      submitter: Party,
+  )(implicit connection: Connection): Set[RejectionReason]
+
+}
+
+object PostCommitValidation {
+
+  /**
+    * Accept unconditionally a transaction.
+    *
+    * Designed to be used by a ledger integration that
+    * already performs post-commit validation.
+    */
+  object Skip extends PostCommitValidation {
+    override def validate(
+        transaction: Transaction,
+        transactionLedgerEffectiveTime: Instant,
+        divulged: Set[ContractId],
+        submitter: Party,
+    )(implicit connection: Connection): Set[RejectionReason] =
+      Set.empty
+  }
+
+  final class BackedBy(data: PostCommitValidationData) extends PostCommitValidation {
+
+    def validate(
+        transaction: Transaction,
+        transactionLedgerEffectiveTime: Instant,
+        divulged: Set[ContractId],
+        submitter: Party,
+    )(implicit connection: Connection): Set[RejectionReason] = {
+
+      val causalMonotonicityRejection =
+        validateCausalMonotonicity(transaction, transactionLedgerEffectiveTime, divulged)
+
+      val invalidKeyUsageRejection =
+        validateKeyUsages(transaction, submitter)
+
+      causalMonotonicityRejection.union(invalidKeyUsageRejection)
+    }
+
+    /**
+      * Do all exercise, fetch and lookup-by-key nodes
+      * 1. exist, and
+      * 2. refer exclusively to contracts with a ledger effective time smaller than or equal to the transaction's?
+      */
+    private def validateCausalMonotonicity(
+        transaction: Transaction,
+        transactionLedgerEffectiveTime: Instant,
+        divulged: Set[ContractId],
+    )(implicit connection: Connection): Set[RejectionReason] = {
+
+      // Collect all the identifiers of contracts that have not been divulged in the current transaction
+      val maximumLedgerEffectiveTime =
+        data.lookupMaximumLedgerTime(
+          transaction.fold(Set.empty[ContractId]) {
+            case (ids, (_, e: Exercise)) if !divulged(e.targetCoid) => ids + e.targetCoid
+            case (ids, (_, f: Fetch)) if !divulged(f.coid) => ids + f.coid
+            case (ids, (_, l: LookupByKey)) => l.result.filterNot(divulged).fold(ids)(ids + _)
+            case (ids, _) => ids
+          }
+        )
+
+      // Query the committed contracts store for the maximum ledger effective time of the
```

Comment:
Should this comment go above line 85? Looks like the querying is happening there.

Reply:
Yes, some code moved and this is a leftover, thanks.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/5737/commits/66a98902caf5384b1d9f95a612b06ccdb7e3f86b